Type-bind for submission input (port from DSpace-CRIS 7 angular)

kshepherd opened the pull request:
## References
* Implements [DS-2873](https://github.com/DSpace/DSpace/issues/2873#issue-659497349)
* Requires DSpace/DSpace#8175

## Description
A port of type-bind functionality ported from DSpace CRIS 7.

The describe forms in submission sections will dynamically update on item type change (publication type, eg. dc.type) to hide or display fields depending on which fields are allowed by the configured type binding in `submission-forms.xml`. As with earlier versions of DSpace, if no type-bind is set for an input field, it will be allowed for all dc.type values.

In addition to the port, the field to check for item type has been made configurable in the environment.submission.* section. Integration / unit testing is WIP.

Original work created by 4Science for DSpace CRIS 7. Ported and modified by The Library Code GmbH with support from Technische Universität Berlin.

## Instructions for Reviewers
Please add a more detailed description of the changes made by your PR. At a minimum, providing a bulleted list of changes in your PR is helpful to reviewers.

List of changes in this PR:
* **DsDynamicTypeBindRelationService** ported, with a few changes to methods and with JSdoc and inline comments added (maybe too much, going by other angular components?)
* **FormBuilderService** updated to handle typeBind model, and some other form / group model handling
* **FieldParser** updated to process typeBind relations on form components to match valid type binds with item type value
* Various dynamic form / field models and components updated to handle typeBind relations
* Some models in tests updated
* Configuration updated to allow setting a different "type" field (default dc.type)

This change is somewhat extensive, touching 23+ files, and comments have been added to try and help with understanding in a code review.

## Checklist

- [ ] My PR is small in size (not quite - ~1200 lines)
- [ ] My PR passes [TSLint](https://palantir.github.io/tslint/) validation using `yarn run lint`
- [ ] My PR doesn't introduce circular dependencies
- [ ] My PR includes [TypeDoc](https://typedoc.org/) comments for _all new (or modified) public methods and classes_. It also includes TypeDoc for large or complex private methods.
- [ ] My PR passes all specs/tests and includes new/updated specs or tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [ ] My PR does NOT include new, third-party dependencies

MW3000 commented:
I was manually testing the functionality, and it works really beautifully. 😎

Only minor issue, I stumbled upon: while rows that are hidden for certain types collapse, columns within a multi-column row stay as an empty space when they are hidden for certain types. It would be nicer if they would also collapse. But this might be a big effort for a small effect. I attach two screenshots to illustrate the issue:

<img width="1182" alt="all-fields-in-row-visible" src="https://user-images.githubusercontent.com/8408850/154519254-b81fb37d-4576-4a6f-afa8-9f7341eb2f34.png">
<img width="1181" alt="only-two-of-three-fiends-in-row-visible" src="https://user-images.githubusercontent.com/8408850/154519293-35256dc1-d473-4a71-97b2-af970e06c36a.png">

atarix83 commented on src/app/shared/form/builder/ds-dynamic-form-ui/models/array-group/dynamic-form-array.component.html:
```diff
@@ -13,15 +14,17 @@
           <!-- Item content -->
-          <div class="drag-handle" [class.drag-disable]="dragDisabled" tabindex="0">
-            <i class="drag-icon fas fa-grip-vertical fa-fw" [class.drag-disable]="dragDisabled" ></i>
+          <div class="drag-handle" [class.invisible]="dragDisabled" tabindex="0">
```
please revert this change otherwise you're reverting a recent fix made here https://github.com/DSpace/dspace-angular/commit/9019b809939868a48d59936ad35b96ab15fcc685

atarix83 reviewed (changes requested):
thanks @kshepherd for this PR

I didn't yet tested but I've put some inline comments to address

kshepherd commented:
Thanks @atarix83 , I will get the securityconfig/level references removed (I thought I had! But a few slipped by) and also revert that template change that you mentioned.

kshepherd commented:
@MW3000 thanks a lot for the feedback! My latest commit should address the layout issue you saw.
It applies the .d-none class to the form-control container in a group or array as well, and I have included a new global css rule to force `visibility: collapse` on ds-form-control-container elements with the .d-none class. This is subject to feedback from more experienced dspace-angular devs as perhaps we should be applying a different class here, or applying the CSS rule in a different place.

But it seems to work well, and you can choose with different bootstrap classes (the `<style>` element in your `submission-forms.xml` field definitions) to choose whether the fields keep their column width or expand to fit the available space.

Example: All 3 fields in my row are displayed as col-sm-4 because type bind = true
![Screenshot from 2022-02-25 12-28-46](https://user-images.githubusercontent.com/194773/155625067-3960ffdc-518d-45ad-ac7d-455d528ba43e.png)

Example: First field is hidden, it collapses and the other fields shift over but retain their width because their class is col-sm-4
![Screenshot from 2022-02-25 12-30-27](https://user-images.githubusercontent.com/194773/155625146-243d0c8c-eeb5-473f-9dff-38bb3cf1850c.png)

Example: First field is hidden, it collapses and the other fields expand to take up available space because I gave them the more simple modern `col-sm` class in the submission form which means bootstrap will just make them even.
![Screenshot from 2022-02-25 12-29-47](https://user-images.githubusercontent.com/194773/155625259-185152c9-7431-455e-b8c8-df63758a1454.png)

(you can also mix and match - perhaps two fields are col-sm-4 but one is col-sm, so if a smaller field is removed, there will be one field taking up 1/3 of the row and another taking up 2/3)

Thanks for the improvement suggestion!

MW3000 commented:
@kshepherd thanks for addressing my feedback.
I tested again, and it works as you describe on Firefox, but unfortunately the changes do not work on Safari. On Safari, it is still the same behaviour as before.

kshepherd commented:
> @kshepherd thanks for addressing my feedback. I tested again, and it works as you describe on Firefox, but unfortunately the changes do not work on Safari. On Safari, it is still the same behaviour as before.

@MW3000 ah, you're right, in Chrome it apparently is also inconsistently implemented... there may not be a perfect way for us to achieve what we want in all browsers as I don't think we can actually destroy the element, but I'll have to look again

kshepherd commented:
@atarix83 I have resolved the issues you identified earlier, and have some basic integration tests included, but I would not consider testing complete as the final behaviour of a form component is not verified.

I think ideally we want to have tests here that can do component fixture detection to see if something is correctly hidden, right? Using the mock submission JSON to put some extra type-bound fields inside... would you recommend I try to do this in `section-form.component.spec.ts`? I might have a few extra questions about the best way to implement tests here.

pnbecker commented:
@MW3000 As it looks to be hard to collapse columns by using CSS, is it acceptable, to solve this by configuration? Can you configure fields without type-bind in columns left from fields using type-bind?

MW3000 commented:
> @MW3000 As it looks to be hard to collapse columns by using CSS, is it acceptable, to solve this by configuration? Can you configure fields without type-bind in columns left from fields using type-bind?

@atarix83 @pnbecker Yes, this is a minor detail and it is totally acceptable to do without it. I was just thoroughly testing ;)

artlowel reviewed (commented):
Thanks @kshepherd!

This is tricky to give feedback on, since you ported code you didn't come up with (and in some cases, according to the comments, even understand 😅) yourself

But I've verified that it works as expected, and put a few inline review comments. As @tdonohue already mentioned: it's mainly lacking tests. 

Since it's a port, it remains to be seen if you can address these issues here yourself, or that they should be addressed upstream, or in a follow up PR by the original creator of the code.

kshepherd commented on src/app/shared/form/builder/ds-dynamic-form-ui/models/date-picker/date-picker.model.ts:
```diff
@@ -25,6 +34,24 @@ export class DynamicDsDatePickerModel extends DynamicDateControlModel {
+        parentModel.hidden = hidden;
+      }
+    });
+  }
```
Thanks, I've tested this with a simple setTimeout() and it is still working well. I see what you mean about the leak potential, I've made a note for the future. Fixed in 74f4384

kshepherd commented on src/app/shared/form/builder/ds-dynamic-form-ui/models/ds-dynamic-row-array-model.ts:
```diff
@@ -21,17 +31,29 @@ export class DynamicRowArrayModel extends DynamicFormArrayModel {
+    }
+    if (isNotUndefined(config.showButtons)) {
+      this.showButtons = config.showButtons;
+    }
```
Thanks, I agree with this. Unless it's decided that false should always be the safe default (and that argument could be made) I've substituted with hasValue. Fixed in 9633a8c

kshepherd commented on src/app/shared/form/builder/form-builder.service.ts:
```diff
@@ -240,6 +281,13 @@ export class FormBuilderService extends DynamicFormService {
+    }
+
+    if (typeBindModel !== null) {
+      this.setTypeBindModel(typeBindModel);
```
Fixed in 9633a8c

artlowel reviewed (approved):
Thanks for addressing my feedback @kshepherd!

As far as I'm concerned this PR can be merged as soon as it passes/has enough tests

kshepherd commented:
Rebased and force pushed as there was a conflict in global css.
I've added service tests but I still have a few things I want to confirm, however any eyes on it would be great.
I ended up with errors in tests with the setTimeout in the date picker model (error calling that from a sync test) and tested it without the async altogether, and it seems to work so it's further simplified.

tdonohue commented on src/app/shared/form/builder/form-builder.service.ts:
```diff
@@ -328,6 +389,82 @@ export class FormBuilderService extends DynamicFormService {
+   * @param id id of model
+   * @param model model
+   */
+  addFormModel(id: string, model: DynamicFormControlModel[]): void {
```
Same here, I cannot seem to find where most of the new methods in `form-builder.service.ts` are actually used.  Are these necessary for this PR, or maybe they were included by accident?

If they are necessary, it might be nice to see if we can find a way to add tests for them.  But, I'm currently not finding how/when they are used.

tdonohue reviewed (commented):
@kshepherd : I gave this a really quick code review today to help this along.  The new tests are looking better, though I added a few suggestions inline.  

I also noticed in reviewing the tests that it looks (to me) like most of the new methods in `form-builder.service.ts` appear to not have tests, but also **appear to not be used anywhere**.  So, rather than asking for more tests, you might see if those methods are necessary...and if not, just remove them from this PR. (Based on my very quick scan of the code, I think they can be removed safely, but it's always possible I've overlooked something.)

@artlowel : wanted to ping you on this PR simply because some new tests are now added.  I know you liked the direction but had asked for tests...so if you find time to give it another review or add in suggestions for testing improvements, that'd be great.

kshepherd commented:
just a note, the 14.x build tests seem to fail sporadically at different places and i can't see that they have anything to do with my changes

kshepherd commented:
@tdonohue @artlowel I've applied feedback regarding tests and fixed remaining LGTM issues. Please let me know if there are any more questions.
Art, I have ensured that the MetadataValidation code in the DSpace PR will not allow values that are not allowed by type binding to make it into workspace items, even if added under a different type, saved by the user, etc., so this should all be taken care of.

tdonohue commented on src/app/shared/empty.util.ts:
```diff
@@ -177,3 +177,29 @@ export const isNotEmptyOperator = () =>
+ * isObjectEmpty({name: null});    // true
+ * isObjectEmpty({ name: 'Adam Hawkins', surname : null});  // false
+ */
+export function isObjectEmpty(obj?: any): boolean {
```
In my previous review I had mentioned adding some basic specs to prove the behavior of this function to the existing `empty.util.spec.ts`.  I still think that'd be worthwhile as it'd prove that it's returning true/false in all the documented scenarios above.

tdonohue reviewed (changes requested):
@kshepherd : Tested this with the backend today. Overall looks great & works well.  I did notice though a few small improvements here.  The main one is that I don't think you need to duplicate the same `typeBind` config on the frontend & backend. See inline below.  Overall though, I think this is looking great.

kshepherd commented on config/config.example.yml:
```diff
@@ -77,6 +77,11 @@ submission:
+    # NOTE: which field to use when matching to type-bind configuration,
+    # eg. dc.type, local.publicationType
+    # default: dc.type
+    field: dc.type
```
@tdonohue the cleanest way i found to do this is to have the config reading in the Form Builder service - this is provided by so many components though, so adding ConfigurationDataService as another dependency means updating a lot of tests and components to provide this, too (or a mock version in the case of tests).
I got about half way through this and wondered if it's worth it... should I instead allow Form Builder to *not* be provided ConfigDataService in which case we'll nicely handle that fact and either return a default "dc.type" type field, or null, or something - but not a fatal error. Or is it better to update all these components? Many of them are not related to submission but in the future may still like to read backend config values or even use type binding.

kshepherd replied:
hm, actually i might have a better way if i play with optional data service, and keep any setting / init out of the constructor so only components that care can provide it

kshepherd replied:
@tdonohue I've add this in, there is now a getTypeField() method in the FormBuilderService. When it is called, and the (optional) config data service is present, the local typeField will subscribe to the configured value from the REST API, and return it. Otherwise it'll return a default of dc_type.
Making the config data service optional was my approach to avoid refactoring many related components and tests that use form builder outside of submission. If it is preferred to wire configuration data service into all the providers that include FormBuilderService instead, I can do this, it's just a bigger change that I wanted to avoid. Happy to hear feedback!

kshepherd replied:
@tdonohue I've added tests for isObjectEmpty, should be good now

tdonohue replied:
Thanks for the update @kshepherd .  I'll give this another review/test once the backend PR is fully updated.  Overall, your approach to resolving this config issue seems fine to me.  I don't have objections.

artlowel reviewed (approved):
I did another code review. 
It looks good 👍

tdonohue reviewed (changes requested):
@kshepherd : Retested the frontend today, and strangely, with the updated changes, I'm finding that this isn't working quite the same.  I *think* it's a frontend issue though, as here is what I'm seeing:

* First, I'm running the frontend in production mode (`yarn build:prod; yarn serve:ssr`)
* On the backend, I've configured one type-bound field in the `traditionalpageone` form (in submission-forms.xml) as follows:
    ```
           <row>
                <field>
                    <dc-schema>dc</dc-schema>
                    <dc-element>identifier</dc-element>
                    <dc-qualifier>isbn</dc-qualifier>
                    <repeatable>false</repeatable>
                    <label>ISBN</label>
                    <input-type>onebox</input-type>
                    <type-bind>Book</type-bind>
                    <hint>Please enter the ISBN of the book.</hint>
                    <required></required>
                </field>
            </row>
    ```
3. In the REST API, I see that the `traditionalpageone` form is returning this field as follows (i.e. it looks like the backend is working, as this looks correct):
    ```
    "fields" : [ {
      "input" : {
        "type" : "onebox"
      },
      "label" : "ISBN",
      "mandatory" : false,
      "repeatable" : false,
      "hints" : "Please enter the ISBN of the book.",
      "selectableMetadata" : [ {
        "metadata" : "dc.identifier.isbn",
        "label" : null,
        "closed" : false
      } ],
      "languageCodes" : [ ],
      "typeBind" : [ "Book" ]
    } ]
    ```
4. However, when I start a new Submission, this type bound field is **immediately visible**, even when the Type field is empty.  Here's a screenshot:
![type-bind](https://user-images.githubusercontent.com/483997/167170656-afbfd85e-48ad-4458-aaf2-bf8711bfb031.png)

I know this all worked the last time I tested it.  So, my best guess is something in the new code may be incorrect & causing this to not work (at least in production mode -- I didn't test in dev mode). 

**UPDATE:**  It appears this problem only happens for the **first new submission**?  Maybe it's a caching issue?  If I discard the first submission & try to submit a second time, everything works (type bind field starts out hidden).  However, if I then go back to the MyDSpace page, hit the *reload* button in the browser & start a new submission, then it **doesn't work**.

kshepherd commented:
@tdonohue thanks, that last issue is fixed, it was introduced with the recent "config from backend" change. Should be good to go now

tdonohue commented on src/app/shared/form/builder/ds-dynamic-form-ui/ds-dynamic-type-bind-relation.service.ts:
```diff
@@ -0,0 +1,230 @@
+  OR_OPERATOR
+} from '@ng-dynamic-forms/core';
+
+import {hasNoValue, hasValue, isEmpty} from '../../../empty.util';
```
According to LGTM, the `isEmpty` here can be removed.  It's not used in this file.

kshepherd replied:
thanks, done

tdonohue reviewed (approved):
👍 Thanks @kshepherd !  Re-tested today and it now works.  So, this looks good to me.

One tiny bit of cleanup... LGTM is still pointing out a single alert/warning on an unused import. I've called it out inline below.

tdonohue commented:
Merging with +2.  Thanks again @kshepherd for your hard work here!  When you get the chance, could you update our Type-Bind docs for DSpace 7 at https://wiki.lyrasis.org/display/DSDOC7x/Submission+User+Interface#SubmissionUserInterface-ItemtypeBasedMetadataCollection ?  Please make sure to note this feature is only available in 7.3 and above, and update those docs as necessary to describe this feature.
